File: service/transport/rmqPublisher.go
package transport

import (
	"log"
	"os"

	"github.com/streadway/amqp"
)

// CreateRmqPublisher create new RmqPublisher
func CreateRmqPublisher(connectionString string) *RmqPublisher {
	return &RmqPublisher{
		connectionString: connectionString,
		logger:           log.New(os.Stdout, "", log.LstdFlags),
	}
}

// RmqPublisher for publish and subscribe
type RmqPublisher struct {
	connectionString string
	logger           *log.Logger
}

// SetLogger set custome logger
func (p *RmqPublisher) SetLogger(logger *log.Logger) Publisher {
	p.logger = logger
	return p
}

// Publish publish message to event
func (p *RmqPublisher) Publish(eventName string, message Message) (err error) {
	// create connection and channel
	conn, ch, err := rmqCreateConnectionAndChannel(p.connectionString)
	if err != nil {
		return err
	}
	defer conn.Close()
	defer ch.Close()
	// declare exchange
	err = rmqDeclareFanoutExchange(ch, eventName)
	if err != nil {
		return err
	}
	// create envelopedMessage
	envelopedMessage, err := NewEnvelopedMessage(message)
	if err != nil {
		return err
	}
	// make json representation of envelopedMessage
	jsonMessage, err := envelopedMessage.ToJSON()
	if err != nil {
		return err
	}
	// publish to exchange
	p.logger.Printf("[INFO RmqPublisher] Publish %s %#v", eventName, message)
	return rmqPublish(ch, eventName, "",
		amqp.Publishing{
			ContentType:   "text/json",
			CorrelationId: envelopedMessage.CorrelationID,
			Body:          jsonMessage,
		})
}

File: templates/python-service/service/transport/rmq_publisher.py
from .interfaces import Message, Publisher
from .helpers import rmq_create_connection_and_channel, rmq_declare_queue_and_bind_to_default_exchange, rmq_publish, rmq_close_connection_and_channel
from .envelopedMessage import EnvelopedMessage
from logging import Logger, getLogger
import json


class RmqPublisher(Publisher):

    def __init__(self, connection_string: str):
        self.connection_string: str = connection_string
        self.logger: Logger = getLogger(__name__.split(".")[0])

    def set_logger(self, logger: Logger) -> Publisher:
        self.logger = logger
        return self

    def publish(self, event_name: str, msg: Message):
        self.logger.info(
            "[INFO RmqPublisher] Publish {} {}".format(event_name, json.dumps(msg)))
        conn, ch = rmq_create_connection_and_channel(self.connection_string)
        rmq_declare_queue_and_bind_to_default_exchange(ch, event_name)
        enveloped_message = EnvelopedMessage().set_correlation_id().set_message(msg)
        rmq_publish(ch, event_name, "", enveloped_message.to_json(), None)
        rmq_close_connection_and_channel(conn, ch)
